myia/anf_ir.py

```diff
@@ -16,11 +16,35 @@
 
 from myia.ir import Node
 from myia.utils import Named
+import json
 
 PARAMETER = Named('PARAMETER')
 APPLY = Named('APPLY')
 
 
+class Debug(types.SimpleNamespace):
+    """Debug information for an object.
+
+    Attributes:
+        name: The name of the object.
+    """
+    __curr_id__ = 0
+
+    def __init__(self, **kw):
+        self.name = None
+        super().__init__(**kw)
+
+    def force_name(self):
```

Comment:
That's what I had at first, but it means we can't check whether a node does have a name or not. `GraphPrinter` needs to know to avoid unnecessarily labelling nodes. Maybe it could be some `has_name()` method.

Reply:
Why would you want to avoid it being triggered?

Reply:
The graphical representation will label the node for `a = x + y` as `add->a`, to signify it's an addition and the result is called `a`. Labelling some anonymous addition `add->#1`, on the other hand, is unnecessary, since that temporary name means nothing for the user and we can draw arrows to signify uses.

Reply:
So you're trying to distinguish between "real" variable names and generated variable names. It sounds to me like that should then just be made explicit i.e. we can have `n.name` which is `a` or `None` and `n.debug_name` which returns `n.name if n.name else generated_name`.

Then, if you *need* a string representation (e.g. in `pdb` or `__repr__`) you use `n.debug_name`, whereas if you only want a name if it corresponds to some semantic variable name (in your visualizations) then you use `n.name`.

Reply:
I mean, it's more subtle than that: *if* the user ever requested to know the name of some node, and `#1` was generated, then when printing out the graphical representation, the corresponding node *should* be labelled `#1`, because the fact it was previously requested means that it is now meaningful. This is starting to feel like bikeshedding though.

Reply:
You can just change `def force_name` into `@property def debug_name`.

Reply:
I kind of prefer `force_name()` because I feel it's a bit more explicit about what it does, but sure.